jumanji/environments/logic/sudoku/utils.py

```diff
@@ -0,0 +1,98 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import chex
+import jax
+import jax.numpy as jnp
+
+from jumanji.environments.logic.sudoku.constants import BOARD_WIDTH, BOX_IDX
+
+
+def apply_action(action: chex.Array, board: chex.Array) -> chex.Array:
+    return board.at[action[0], action[1]].set(action[2])
+
+
+def puzzle_completed(board: chex.Array) -> chex.Array:
+    """Checks that every row, column and 3x3 boxes includes all figures.
+
+    Args:
+        board: The sudoku board.
+
+    Returns:
+        condition: A `bool` indicator that validates if the puzzle is solved.
+    """
+
+    def _validate_row(row: chex.Array) -> chex.Array:
+        condition = jnp.sort(row) == jnp.arange(BOARD_WIDTH)
```

Comment:
Wouldn't using jnp.unique be (slightly) faster?

Reply:
Given the size of the array (9) I don't think it really matters, but anyway I run a small comparison on my CPU.
```python
import jax.numpy as jnp
import jax
​
a = jnp.array([1,3,5,2,0,0,7,6,0])
​
sort_fn = jax.jit(jnp.sort)
unique_fn = jax.jit(lambda x: jnp.unique(x, size=9))
​
warmup_sort = sort_fn(a).block_until_ready()
warmup_unique = unique_fn(a).block_until_ready()
​
%timeit sort_fn(a).block_until_ready()
%timeit unique_fn(a).block_until_ready()
2.43 µs ± 13.1 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
2.96 µs ± 47.5 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```